# Fundamentals Of Data Analysis - Tasks

**Kevin O'Leary**

***

## Task 1 - Collatz Function

>Verify, using Python, that the Collatz conjecture is true for the first 10,000 positive integers.

In [ ]:
def f(x):
  #If x is even, divide it by two.
  if x % 2 == 0:
    return x // 2
  else:
    return (3 * x) + 1

In [ ]:
conjecture_success = []  # If the conjecture works "True" will be added to this list. If it fails "False" will. This way we  
                         # can check it works without printing out all sequences. 

In [ ]:
def collatz(x):
  while x != 1:          # if we run x as 1 it will get caught in a loop. we want it to stop when x = 1,   
                         # as the conjecture is then true for that value
    
    #print(x, end=', ')  # this would print out each digit in the sequence for each value which is un-neccessary.  
                         # We just need to know if its true.
    x=f(x)
  
  if x == 1: 
    conjecture_success.append("True")     # if the last value of x is 1, then it conforms to the collatz conjecture  
                                          # and the result "True" is added to our list. 
  elif x != 1:       
    conjecture_success.append("False")    # if the last value of x isnt 1, then it has failed the collatz conjecture  
                                          # and the result "False" is added to our list.
      
  

In [ ]:
x = 1
while x <= 10000:  #  a while loop running the collatz function for all values 1-10,000
  collatz(x)
  x += 1


In [ ]:
if "False" in conjecture_success:
  print("The conjecture has been disproven")
else:
  print("Conjecture is true for all values up to 10,000")


## Task 2 - Penguin Dataset 

>Summary of the Penguin Database and suggest variables that could be used to model it

**Summary**

The penguins dataset was collected by Dr Kristen Gorman at the Palmer Station in Antartica. it contains information about different species of penguins at the Palmer Archipeligo. It is commonly used to demonstrate various data analyses and machine learning techniques. 

In [1]:
import seaborn as sb
#Seaborn is a Data Visualisation Library. It contains a number of datasets including penguins.csv. 

In [2]:
import pandas as pd


In [3]:
sb.get_dataset_names()

# Returns a list of all datasets within Seaborn

['anagrams',
 'anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'dowjones',
 'exercise',
 'flights',
 'fmri',
 'geyser',
 'glue',
 'healthexp',
 'iris',
 'mpg',
 'penguins',
 'planets',
 'seaice',
 'taxis',
 'tips',
 'titanic']

In [4]:
df = sb.load_dataset('penguins')

# loads the penguins dataset. 
# ref: https://seaborn.pydata.org/tutorial/introduction.html

In [5]:
df.shape

# displays how many rows/columns there are.

(344, 7)

The dataframe is made up of 7 columns and 344 rows

In [6]:
df.head()

# Initial look at the dataset to ensure its what we were looking for. The head function allows you to see the first 5 rows in the dataframe

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

From looking at the above we can see the dataframe is made up for 7 columns of different variables. Species, Island of Origin, Bill Length (mm), Bill Depth (mm), Flipper Length (mm), body mass (g) and Sex

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
dtypes: float64(4), object(3)
memory usage: 18.9+ KB


In [10]:
df.dtypes

species               object
island                object
bill_length_mm       float64
bill_depth_mm        float64
flipper_length_mm    float64
body_mass_g          float64
sex                   object
dtype: object

Pandas classification of datatypes is a little different to Python. In Python we have the datatypes; int, float, string and bool. However in Pandas int is classified as 'int64', float as 'float64', a string as 'object' and bool remains the same. 

Based on the above, 'species', 'island' and 'sex are strings in Python and object in pandas. The other columns that contain measurements area all floats in Python and float64 in Pandas. This is what we would expect based on the contents of the columns above. 

Ref: https://pbpython.com/pandas_dtypes.html

***

## End